src/nlp.jl

```diff
@@ -83,147 +201,87 @@ function MathProgBase.initialize(d::JuMPNLPEvaluator, requested_features::Vector
+    if :ExprGraph in requested_features
+        d.subexpressions_as_julia_expressions = Array(Any,length(subexpr))
+        for k in d.subexpression_order
+            ex = nldata.nlexpr[k]
+            adj = adjmat(ex.nd)
+            d.subexpressions_as_julia_expressions[k] = tapeToExpr(1, ex.nd, adj, ex.const_values, d.parameter_values, d.subexpressions_as_julia_expressions)
         end
-        rowoffset = size(A,1)+1
-        for c::QuadConstraint in d.m.quadconstr
-            aff = c.terms.aff
-            for k in 1:length(aff.vars)
-                jac_I[idx] = rowoffset
-                jac_J[idx] = aff.vars[k].col
-                idx += 1
-            end
-            for k in 1:length(c.terms.qvars1)
-                jac_I[idx] = rowoffset
-                jac_J[idx] = c.terms.qvars1[k].col
-                jac_I[idx+1] = rowoffset
-                jac_J[idx+1] = c.terms.qvars2[k].col
-                idx += 2
+    end
+
+    subexpression_linearity = Array(Linearity, length(nldata.nlexpr))
+    subexpression_variables = Array(Set{Int}, length(nldata.nlexpr))
+    subexpression_edgelist = Array(Set{Tuple{Int,Int}}, length(nldata.nlexpr))
+    d.subexpressions = Array(SubexpressionStorage, length(nldata.nlexpr))
+    for k in d.subexpression_order # only load expressions which actually are used
+        d.subexpressions[k] = SubexpressionStorage(nldata.nlexpr[k], numVar, want_hess_storage)
+        subex = d.subexpressions[k]
+        if d.want_hess
+            linearity = classify_linearity(subex.nd, subex.adj, subexpression_linearity)
+            subexpression_linearity[k] = linearity[1]
+            vars = compute_gradient_sparsity(d.subexpressions[k].nd)
+            # union with all dependent expressions
+            for idx in list_subexpressions(d.subexpressions[k].nd)
+                union!(vars, subexpression_variables[idx])
             end
-            rowoffset += 1
+            subexpression_variables[k] = vars
+            edgelist = compute_hessian_sparsity(subex.nd, subex.adj, linearity,subexpression_edgelist, subexpression_variables)
+            subexpression_edgelist[k] = edgelist
         end
-        for k in 1:length(nljacI)
-            jac_I[idx] = nljacI[k]+rowoffset-1
-            jac_J[idx] = nljacJ[k]
-            idx += 1
-        end
-        @assert idx-1 == nnz_jac
+
     end
-    d.jac_I = jac_I
-    d.jac_J = jac_J
 
-    if need_hessian
-        if d.has_nlobj
-            hI, hJ, d.eval_h_nl = gen_hessian_sparse_color_parametric(nldata.nlobj, d.m.numCols)
-            nnz_hess += length(hI)
-        else
-            hI = []
-            hJ = []
-            d.eval_h_nl = (x,y,z) -> nothing
-            nnz_hess += length(d.m.obj.qvars1)
-        end
-        d.nnz_hess_obj = length(hI)
 
-        hess_I = zeros(Int,nnz_hess)
-        hess_J = zeros(Int,nnz_hess)
+    if d.has_nlobj
+        @assert length(d.m.obj.qvars1) == 0 && length(d.m.obj.aff.vars) == 0
```

Comment:
Just for maintainability, it might be good to break this into two lines (separate assertions) so that it is clear which length is incorrect should either trigger an error.